File: Prototypes/SensorDetector/app/src/main/java/com/example/sensordetector/MainActivity.java
package com.example.sensordetector;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import java.util.Arrays;
import java.util.Date;

import org.w3c.dom.Text;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = "MainActivity";
    private SensorManager sensorManager;
    private Sensor sensor;
    private TextView txt_xAxis;
    private Button btnStart, btnStop;

    private double[][] gesture_matrix = new double[100][];
    private int matrix_idx = 0;

    private SensorEventListener mSensorListener = new SensorEventListener() {
        @Override
        public void onSensorChanged(SensorEvent sensorEvent) {
            double[] values = {sensorEvent.values[0],sensorEvent.values[1],sensorEvent.values[2]};

            if (matrix_idx < 100){
                gesture_matrix[matrix_idx] = values;
                matrix_idx += 1;
            }
        }

        @Override
        public void onAccuracyChanged(Sensor sensor, int i) {

        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState){
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        txt_xAxis = (TextView) findViewById(R.id.xAxis);
        btnStart = (Button) findViewById(R.id.btn_start);
        btnStop = (Button) findViewById(R.id.btn_stop);

        Log.d(TAG, "onCreate: Initializing Sensor Services");
        sensorManager = (SensorManager) getSystemService(Context.SENSOR_SERVICE);

        btnStart.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startGesture();
            }
        });

        btnStop.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                endGesture();
            }
        });

        sensor = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER);
        Log.d(TAG, "onCreate: Registered Accelerometer listener");
    }

    private void startGesture() {
        btnStart.setEnabled(false);
        btnStop.setEnabled(true);

        sensorManager.registerListener(mSensorListener, sensor, SensorManager.SENSOR_DELAY_NORMAL);
    }

    private void endGesture() {
        sensorManager.unregisterListener(mSensorListener);
        btnStart.setEnabled(true);
        btnStop.setEnabled(false);

        txt_xAxis.setText(Arrays.deepToString(gesture_matrix));
        Log.d(TAG, "Gesture Matrix Data: " + Arrays.deepToString(gesture_matrix));
    }

}

File: Prototypes/CustomModelSample/app/src/main/java/com/example/custommodelsample/MainActivity.kt
package com.example.custommodelsample

import android.app.Activity
import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Bundle
import com.example.custommodelsample.databinding.ActivityMainBinding
import java.nio.ByteBuffer

import android.support.v4.app.RemoteActionCompatParcelizer
import android.widget.Button
import android.widget.Toast
import com.example.custommodelsample.ml.GestureModel
import org.tensorflow.lite.DataType
import org.tensorflow.lite.support.tensorbuffer.TensorBuffer

class MainActivity : Activity() {
    //val initializeTask: Task<Void> by lazy { TfLite.initialize(this) }

    private lateinit var binding: ActivityMainBinding
    //private lateinit var interpreter: InterpreterApi
    private lateinit var byteBuffer: ByteBuffer;
    private val input = FloatArray(127*3)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val sensorManager = getSystemService(Context.SENSOR_SERVICE) as SensorManager
        val accelerometer: Sensor? = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)

        /*
        val gyroscope: Sensor? = sensorManager.getDefaultSensor(Sensor.TYPE_GYROSCOPE)
        sensorManager.registerListener(GyroscopeListener(fun(x: Float,y:Float,z:Float){
            binding.gyroscopeText.text = "Gyroscope Readings $x $y $x"
        }), gyroscope, SensorManager.SENSOR_DELAY_NORMAL);
         */


        binding.btnStart.setOnClickListener{
            sensorManager.registerListener(AccelerometerListener(fun(x: Float,y:Float,z:Float){
                for (i in 0..380 step 3){
                    input[i] = x
                    input[i+1] = y
                    input[i+2] = z
                }
                //binding.accelerometerText.text = "Accelerometer Readings $x $y $x"
            }), accelerometer, SensorManager.SENSOR_DELAY_NORMAL);
        }

        binding.btnStop.setOnClickListener {
            sensorManager.unregisterListener(AccelerometerListener(fun(x: Float,y:Float,z:Float){}))

            val prediction = Predict(input)

            if(prediction[0] > prediction[1]){
                binding.txtPrediction.text = "Prediction: Right"
            }else if(prediction[0] < prediction[1]){
                binding.txtPrediction.text = "Prediction: Left"
            }else{
                binding.txtPrediction.text = "Prediction: Neither"
            }
        }
    }

    fun Predict(value: FloatArray) : FloatArray{
        val model = GestureModel.newInstance(this)
        // Creates inputs for reference.
        val inputFeature0 = TensorBuffer.createFixedSize(intArrayOf(1, 127, 3), DataType.FLOAT32)
        //inputArray = value
        inputFeature0.loadArray(value)
        // Runs model inference and gets result.
        val outputs = model.process(inputFeature0)
        val outputFeature0 = outputs.outputFeature0AsTensorBuffer

        val predictedValue = floatArrayOf(outputFeature0.getFloatValue(0), outputFeature0.getFloatValue(1))

        // Releases model resources if no longer used.
        model.close()
        return predictedValue
    }

    private val sensorEventListener: SensorEventListener = object : SensorEventListener {
        override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {}
        override fun onSensorChanged(event: SensorEvent) {

        }
    }
}

enum class DetectorState {
    Idle,
    Recording,
    Detecting,
}

private class Detector{
    var currentState: DetectorState = DetectorState.Idle;

    fun Idle(){

    }

    fun Recording(){
        currentState = DetectorState.Recording;
        var recordingState = RecordingState()

    }

    fun Detecting(){
        currentState = DetectorState.Detecting;
    }
}

private class RecordingState{
    fun StartRecording(){

    }

    fun StopRecording(){

    }
}

private class DetectingState(){
    fun Detect(){

    }
}


private class AccelerometerListener(val callback: (Float, Float, Float) -> Unit)  : SensorEventListener{

    override fun onSensorChanged(event: SensorEvent) {
        val x = event.values[0];
        val y = event.values[1];
        val z = event.values[2];

        callback(x,y,z)
    }

    override fun onAccuracyChanged(p0: Sensor?, p1: Int) {
    }
}

private class GyroscopeListener(val callback: (Float, Float, Float) -> Unit) : SensorEventListener{
    override fun onSensorChanged(event: SensorEvent) {
        val x = event.values[0];
        val y = event.values[1];
        val z = event.values[2];

        callback(x,y,z)
    }

    override fun onAccuracyChanged(p0: Sensor?, p1: Int) {
    }
}